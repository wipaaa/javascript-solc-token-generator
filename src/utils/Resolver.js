const fs = require('fs');
const path = require('path');
const NotFoundError = require('../exceptions/NotFoundError');

module.exports = class Resolver {
  constructor() {
    this.sources = [];
    this.defaults = {
      charset: 'utf-8',
      data: {
        name: null,
        content: null,
        path: null,
      },
      directory: '../../stub',
      extension: {
        from: '.txt',
        to: '.sol',
      },
    };
  }

  #_getActualPath = (...pathSegments) => path.resolve('src', ...pathSegments);

  #_getFormattedFilename = (filename) => {
    const { from, to } = this.defaults.extension;
    return filename.replace(from, to);
  };

  #_isNotExists = (pathTo) => !fs.existsSync(this.#_getActualPath(pathTo));

  #_mergeWithDefaultData = (newData) => ({ ...this.defaults.data, ...newData });

  #_pullSources = () => {
    if (!this.sources.length) {
      return this.sources;
    }

    const sources = [...this.sources];
    this.sources = [];

    return sources;
  };

  #_readContentFrom = (pathToFile) => {
    const { charset } = this.defaults;
    const actualPath = this.#_getActualPath(pathToFile);

    return fs.readFileSync(actualPath, charset).toString();
  };

  #_readFilesFrom = (pathToDirectory, callback) => {
    const { charset } = this.defaults;
    const actualPath = this.#_getActualPath(pathToDirectory);

    return fs.readdirSync(actualPath, charset).map(callback);
  };

  #_resolveFromFile = (pathToFile) => {
    if (this.#_isNotExists(pathToFile)) {
      throw new NotFoundError(`File isn't exists at: ${pathToFile}`);
    }

    const name = this.#_getFormattedFilename(path.basename(pathToFile));
    const content = this.#_readContentFrom(pathToFile);

    return this.#_mergeWithDefaultData({
      name,
      content,
      path: path.dirname(pathToFile),
    });
  };

  #_resolveFromSources = () => {
    if (!this.sources.length) {
      return null;
    }

    this.sources.forEach((source, id) => {
      const content = this.#_readContentFrom(source.path);
      this.sources[id] = { ...source, content };
    });

    return null;
  };
};
