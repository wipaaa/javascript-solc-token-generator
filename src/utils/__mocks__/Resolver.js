const fs = require('fs');
const path = require('path');

const _getActualPath = (pathTo) => {
  return path.resolve('src', pathTo);
};

module.exports = class Resolver {
  constructor() {
    this.sources = [];
    this.defaults = {
      charset: 'utf-8',
      data: {
        name: null,
        content: null,
        path: null,
      },
      directory: '../../stub',
      extension: {
        from: '.txt',
        to: '.sol',
      },
    };
  }

  resolve = jest.fn().mockImplementation(() => {
    const sources = this.sources;
    this.sources = [];

    return sources;
  });

  setFrom = jest.fn().mockImplementation((pathToDirectory) => {
    const actualPath = _getActualPath(pathToDirectory);

    if (!fs.existsSync(actualPath)) {
      const message = `Directory isn't exists at: ${pathToDirectory}.`;
      throw new Error(message);
    }

    this.sources = fs.readdirSync(actualPath).map((source) => {
      const { extension } = this.defaults;

      return {
        name: source.replace(extension.from, extension.to),
        content: null,
        path: actualPath,
      };
    });

    return this;
  });
};
