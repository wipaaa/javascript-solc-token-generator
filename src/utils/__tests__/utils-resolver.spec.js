import Resolver from '../Resolver';

describe('Resolver utility test', () => {
  let resolver = null;

  beforeAll(() => {
    resolver = new Resolver();
  });

  describe('when instantiated', () => {
    it('should be an instance of resolver class', () => {
      expect(resolver).toBeTruthy();
      expect(resolver).toBeInstanceOf(Resolver);
    });

    it('should have property sources', () => {
      expect(resolver).toHaveProperty('sources', []);
    });

    it('should have property defaults and match defaults property snapshot', () => {
      expect(resolver).toHaveProperty('defaults');
      expect(resolver.defaults).toMatchSnapshot();
    });
  });

  describe('when resolve method called', () => {
    beforeEach(() => {
      resolver.sources = [];
    });

    it('should not throw any error', () => {
      expect(() => resolver.resolve()).not.toThrow();
    });

    it('should return empty array when no sources provided', () => {
      const sources = resolver.resolve();
      expect(sources).toEqual([]);
    });

    it('should match the sources snapshot', () => {
      const path = '../stub';
      const sources = resolver.setFrom(path).resolve();
      expect(sources).toMatchSnapshot();
    });
  });

  describe('when setFrom method called', () => {
    beforeEach(() => {
      resolver.sources = [];
    });

    it('should throw an error while provided with invalid path', () => {
      expect(() => resolver.setFrom('/fake/path')).toThrow();
    });

    it('should return the resolver object itself', () => {
      const instance = resolver.setFrom('../stub');

      expect(instance).toBeTruthy();
      expect(instance).toBeInstanceOf(Resolver);
    });

    it('should change the sources property', () => {
      resolver.setFrom('../stub');
      expect(resolver.sources).not.toEqual([]);
    });

    it('should match the sources snapshot', () => {
      resolver.setFrom('../stub');
      expect(resolver.sources).toMatchSnapshot();
    });
  });
});
