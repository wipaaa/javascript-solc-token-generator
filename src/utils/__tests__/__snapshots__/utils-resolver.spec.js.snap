// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Resolver utility when resolve method called should return the sources object 1`] = `
Array [
  Object {
    "content": null,
    "name": "Context.sol",
    "path": "/home/wipaaa/github/javascript-solc-token-generator/stub",
  },
  Object {
    "content": null,
    "name": "Ownable.sol",
    "path": "/home/wipaaa/github/javascript-solc-token-generator/stub",
  },
  Object {
    "content": null,
    "name": "SafeMath.sol",
    "path": "/home/wipaaa/github/javascript-solc-token-generator/stub",
  },
  Object {
    "content": null,
    "name": "Token.sol",
    "path": "/home/wipaaa/github/javascript-solc-token-generator/stub",
  },
]
`;

exports[`Resolver utility when setFrom method called should change the sources property with directory files list 1`] = `
Array [
  Object {
    "content": null,
    "name": "Context.sol",
    "path": "/home/wipaaa/github/javascript-solc-token-generator/stub",
  },
  Object {
    "content": null,
    "name": "Ownable.sol",
    "path": "/home/wipaaa/github/javascript-solc-token-generator/stub",
  },
  Object {
    "content": null,
    "name": "SafeMath.sol",
    "path": "/home/wipaaa/github/javascript-solc-token-generator/stub",
  },
  Object {
    "content": null,
    "name": "Token.sol",
    "path": "/home/wipaaa/github/javascript-solc-token-generator/stub",
  },
]
`;
